torch/csrc/distributed/c10d/reducer.cpp

```diff
@@ -442,12 +442,19 @@ void Reducer::mark_variable_ready_sparse(size_t variable_index) {
   });
 }
 
-std::vector<at::Tensor> Reducer::get_bucket_tensors() const {
+std::vector<c10d::GradBucket> Reducer::get_bucket_tensors() const {
   std::lock_guard<std::mutex> lock(mutex_);
-  std::vector<at::Tensor> bucketTensors;
+  std::vector<c10d::GradBucket> bucketTensors;
   bucketTensors.reserve(buckets_.size());
-  for (const auto& bucket : buckets_) {
-    bucketTensors.emplace_back(bucket.replicas[0].contents);
+  for (int i = 0; i < buckets_.size(); ++i) {
+    bucketTensors.emplace_back(
+      i,
+      // This API is only used for uneven inputs which requires zero tensors
+      at::zeros_like(buckets_[i].replicas[0].contents),
```

Comment:
I think this can surprise users.  According to the function name, I won't expect any transformation after fetching the data.

What about moving ``zeros_like` to L1024 of file `distributed.py`?

Reply:
Yeah, agreed it is not ideal. However we can't really do the `zeros_like` in python anymore because we cannot construct a new GradBucket and there's no API to change the tensor corresponding to a `GradBucket` (the latter can probably be misused).

If we really don't want this, we can add back being able to construct a `GradBucket` in python (which I removed recently), although this will require constructing an extra `GradBucket` when we really don't need to.

Reply:
What about providing 3 functions and only expose 1 as a public API for handling uneven inputs?
1. get_grad_buckets w/ this flag
2. get_grad_buckets w/o this flag, which calls 1)
3. get_zeros_like_grad_buckets w/o this flag, which calls 1), and only expose this as a public API.

Reply:
Updated based on this suggestion, PTAL